docs/docs/how-tos/multi-agent-multi-turn-convo-functional.ipynb

```diff
@@ -0,0 +1,430 @@
+    "from langchain_core.messages import AIMessage\n",
+    "from langchain_anthropic import ChatAnthropic\n",
+    "from langgraph.prebuilt import create_react_agent\n",
+    "from langgraph.graph import add_messages\n",
+    "from langgraph.func import entrypoint, task\n",
+    "from langgraph.checkpoint.memory import MemorySaver\n",
+    "from langgraph.types import interrupt, Command\n",
+    "\n",
+    "model = ChatAnthropic(model=\"claude-3-5-sonnet-latest\")\n",
+    "\n",
+    "# Define travel advisor ReAct agent\n",
+    "travel_advisor_tools = [\n",
+    "    get_travel_recommendations,\n",
+    "    transfer_to_hotel_advisor,\n",
+    "]\n",
+    "travel_advisor = create_react_agent(\n",
+    "    model,\n",
+    "    travel_advisor_tools,\n",
+    "    state_modifier=(\n",
+    "        \"You are a general travel expert that can recommend travel destinations (e.g. countries, cities, etc). \"\n",
+    "        \"If you need hotel recommendations, ask 'hotel_advisor' for help. \"\n",
+    "        \"You MUST include human-readable response before transferring to another agent.\"\n",
+    "    ),\n",
+    ")\n",
+    "\n",
+    "\n",
+    "@task\n",
+    "def call_travel_advisor(messages):\n",
+    "    # You can also add additional logic like changing the input to the agent / output from the agent, etc.\n",
+    "    # NOTE: we're invoking the ReAct agent with the full history of messages in the state\n",
+    "    response = travel_advisor.invoke({\"messages\": messages})\n",
+    "    return response[\"messages\"]\n",
+    "\n",
+    "\n",
+    "# Define hotel advisor ReAct agent\n",
+    "hotel_advisor_tools = [get_hotel_recommendations, transfer_to_travel_advisor]\n",
+    "hotel_advisor = create_react_agent(\n",
+    "    model,\n",
+    "    hotel_advisor_tools,\n",
+    "    state_modifier=(\n",
+    "        \"You are a hotel expert that can provide hotel recommendations for a given destination. \"\n",
+    "        \"If you need help picking travel destinations, ask 'travel_advisor' for help.\"\n",
+    "        \"You MUST include human-readable response before transferring to another agent.\"\n",
+    "    ),\n",
+    ")\n",
+    "\n",
+    "\n",
+    "@task\n",
+    "def call_hotel_advisor(messages):\n",
+    "    response = hotel_advisor.invoke({\"messages\": messages})\n",
+    "    return response[\"messages\"]\n",
+    "\n",
+    "\n",
+    "checkpointer = MemorySaver()\n",
+    "\n",
+    "\n",
+    "@entrypoint(checkpointer=checkpointer)\n",
+    "def multi_turn_graph(messages, previous):\n",
+    "    previous = previous or []\n",
+    "    messages = add_messages(previous, messages)\n",
+    "\n",
+    "    call_active_agent = call_travel_advisor\n",
+    "    while True:\n",
+    "        agent_messages = call_active_agent(messages).result()\n",
+    "        messages = add_messages(messages, agent_messages)\n",
+    "        ai_msg = next(m for m in reversed(agent_messages) if isinstance(m, AIMessage))\n",
```

Comment:
This was extremely confusing for me, I added the following comment:

```
    // Find the last AI message
    // If one of the handoff tools is called, the last message returned
    // by the agent will be a ToolMessage because we set them to have
    // "returnDirect: true". This means that the last AIMessage will
    // have tool calls.
    // Otherwise, the last returned message will be an AIMessage with
    // no tool calls, which means we are ready for new input.
```